libcfnet/conn_cache.c

```diff
@@ -128,6 +128,22 @@ AgentConnection *ConnCache_FindIdleMarkBusy(const char *server,
             {
                 assert(svp->status == CONNCACHE_STATUS_IDLE);
 
+                // Check connection state before returning it
+                int error = 0;
+                socklen_t len = sizeof (error);
+                if (getsockopt(svp->conn->conn_info->sd, SOL_SOCKET, SO_ERROR, &error, &len) < 0) {
+                    Log(LOG_LEVEL_VERBOSE, "FindIdle:"
+                        " found connection to '%s' but could not get socket status, skipping.",
+                        server);
+                    continue;
+                }
+                if (error) {
+                    Log(LOG_LEVEL_VERBOSE, "FindIdle:"
+                        " found connection to '%s' but connection is broken, skipping.",
+                        server);
+                    continue;
+                }
```

Comment:
Shouldn't both these cases result into the connection being removed from the cache? What's the point in trying it again and again in later attempts?

Reply:
Mainly because I did not know how to simple handle this from inside the `for (size_t i = 0; i < SeqLength(conn_cache); i++)` loop without messing with the iteration.

Reply:
I think we need a flag in the `ConnCache_entry` struct to mark connections as invalid to either skip them quickly/cheaply or remove them after the loop. Will you add that @amousset or shall I add it as a separate PR once this is merged?

Reply:
I won't have much time to work on this in the following weeks, so it would probably be more efficient if you do it in a separate PR :)

Reply:
> I won't have much time to work on this in the following weeks, so it would probably be more efficient if you do it in a separate PR :)

Works for me.